### PR TITLE
BUG: Prevent invalid array shapes in seed

### LINKED ISSUE #9832: numpy.random.seed fails with empty non-int array
```Python
>>> numpy.random.seed([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "mtrand.pyx", line 684, in mtrand.RandomState.seed
TypeError: Cannot cast array from dtype('float64') to dtype('int64') according to the rule 'safe'
```
```Python
>>> numpy.random.seed(numpy.array([]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "mtrand.pyx", line 684, in mtrand.RandomState.seed
TypeError: Cannot cast array from dtype('float64') to dtype('int64') according to the rule 'safe'
```

### PR DESCRIPTION
**Pull request by bashtage**
Prevent empty arrays or arrays with more than 1 dimension from being
used to seed RandomState

closes #9832

### CODE REVIEW

**Comment by bashtage** [comment]
The extra `asarray` was added to catch empty arrays.  If this wasn't added, then a `TypeError` was raised about float conversion and it wasn't possible to give a clean explanation.

**Comment by rkern** [comment]
It seems like the second `np.asarray(obj)` is redundant, though. Other than that, LGTM.

**Comment by bashtage** [comment]
> It seems like the second np.asarray(obj) is redundant, though. Other than that, LGTM.

Hopefully ready now.

**Comment by rkern** [comment]
LGTM!

File: numpy/random/mtrand/mtrand.pyx [hunk 1]
```diff
@@ -681,9 +681,14 @@ cdef class RandomState:
                 with self.lock:
                     rk_seed(idx, self.internal_state)
         except TypeError:
-            obj = np.asarray(seed).astype(np.int64, casting='safe')
-            if ((obj > int(2**32 - 1)) | (obj < 0)).any():
-                raise ValueError("Seed must be between 0 and 2**32 - 1")
+            obj = np.asarray(seed)
+            if obj.size == 0:
+                raise ValueError("Seed must convertible to a 32-bit unsigned "
+                                 "integer and non-empty")
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: "convertible to a 32-bit unsigned integer" doesn't seem relevant to the problem here

File: numpy/random/mtrand/mtrand.pyx [hunk 2]
```diff
@@ -681,9 +681,14 @@ cdef class RandomState:
...
+            obj = obj.astype(np.int64, casting='safe')
+            if ((obj > int(2**32 - 1)) | (obj < 0)).any() or obj.ndim != 1:
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: Any reason to not split this into two tests?

Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: `>= 2**32` would probably be clearer here, while you're touching this line.

File: numpy/random/tests/test_random.py [hunk 3]
```diff
@@ -42,6 +42,13 @@ def test_invalid_array(self):
         assert_raises(ValueError, np.random.RandomState, [1, 2, 4294967296])
         assert_raises(ValueError, np.random.RandomState, [1, -2, 4294967296])
 
+    def test_invalid_array_shape(self):
+        # GH 9832
```
Comment (eric-wieser) [inline numpy/random/tests/test_random.py]: Elsewhere I think we use `gh-9832`, ie lower-case and with a hyphen

File: numpy/random/mtrand/mtrand.pyx (hunk 1 above)
Comment (rkern) [reply numpy/random/mtrand/mtrand.pyx]: It's just a way of restating the original message and emphasizing that the values must be integral rather than just floats in that range.

**Comment by bashtage** [comment]
I have simplified these a bit and I think they match the error detected a bit better.

File: numpy/random/mtrand/mtrand.pyx [hunk 4]
```diff
@@ -681,8 +681,13 @@ cdef class RandomState:
...
         except TypeError:
-            obj = np.asarray(seed).astype(np.int64, casting='safe')
-            if ((obj > int(2**32 - 1)) | (obj < 0)).any():
+            obj = np.asarray(seed)
+            if obj.size == 0:
+                raise ValueError("Seed must be non-empty")
+            obj = obj.astype(np.int64, casting='safe')
+            if obj.ndim != 1:
+                raise ValueError("Seed array must be 1-d")
+            if ((obj >= int(2**32)) | (obj < 0)).any():
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: I know it was here before, but this `int` cast is pointless

**Comment by bashtage** [comment]
I assume is was used on 32 bit OSes to push the 2**32-1 back from a Python
long to a Python int on 2.x.   I agree it is pointless now.

File: numpy/random/mtrand/mtrand.pyx [hunk 5]
```diff
@@ -681,8 +681,13 @@ cdef class RandomState:
...
         except TypeError:
-            obj = np.asarray(seed).astype(np.int64, casting='safe')
-            if ((obj > int(2**32 - 1)) | (obj < 0)).any():
+            obj = np.asarray(seed)
+            if obj.size == 0:
+                raise ValueError("Seed must be non-empty")
+            obj = obj.astype(np.int64, casting='safe')
+            if obj.ndim != 1:
+                raise ValueError("Seed array must be 1-d")
+            if ((obj >= 2**32) | (obj < 0)).any():
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: Sorry, just spotted that making this `2**32` makes this inconsistent with the `if` above - can you change the other one too? (or change this one back)

**Comment by charris** [comment]
Looks easy to finish this up.

**Comment by bashtage** [comment]
Ready.
